src/utils/simulation.js

```diff
@@ -0,0 +1,210 @@
+import generateRandomnArray from "src/utils/generateRandomArray";
+import logger from "src/utils/logger";
+
+/** Class representing a simulation. */
+class Simulation {
+  /**
+   * @param {number} pop_size - The population size.
+   * @param {number} vacc_percentage - The percentage of population vaccinated.
+   * @param {number} initial_infected - The percentage of population vaccinated. 
+   * @param {object} virus - The virus object.
+   * @param {number} total_interactions - The total interactions every infected person will complete.
+   */
+  constructor(pop_size, vacc_percentage, initial_infected=0, virus, total_interactions) {
+    this.pop_size = pop_size;
+    this.vacc_percentage = vacc_percentage;
+    this.initial_infected = initial_infected;
+    this.virus = virus;
+    this.total_interactions = total_interactions;
+    this.logger = new logger();
+    this.newly_infected = [];
+    this.current_infected = [];
+    this.population = this.create_population();
+  }
+
+/**
+ * @param {number} id - The id of the current person infected.
+ * It filters through population for people who can still get sick.  
+ * @returns {array} - People who are not vaccinated and infected.
+ */
+  alivePeople(id) {
+    return this.population.filter((person) => person.is_alive && id !== person._id && !person.is_vaccinated && !person.infection);
+  }
+
+  /**
+   * Creates the initial population.
+   * People are randomly chosen to be initially sick and vaccinated according to given parameters.
+   * Infected people's ids are added to current_infected empty array.
+   * @return {array} pop - A list of Person objects.
+   */
+  
+  create_population() {
+    // create people objects
+    let pop = [];  
+    for (let i = 0; i < this.pop_size; i++) {
+      pop.push(new Person(i, false))
+    }
+
+    // calculate initial vaccination rate
+    const initial_vacc = Math.ceil(this.pop_size * this.vacc_percentage);
+    const randomPeopleIds = generateRandomnArray((this.initial_infected + initial_vacc),this.pop_size);
+
+    // infect people
+    for (let i = 0; i < this.initial_infected; i++) {
+      const random_person = randomPeopleIds[i]; 
+      pop[random_person].infection = this.virus;
+      this.current_infected.push(randomPeopleIds[i]);
+    }
+
+    // vaccinate people
+    for (let i = this.initial_infected; i < randomPeopleIds.length; i++) {
+      const random_person = randomPeopleIds[i];
+      pop[random_person].is_vaccinated = true; 
+    }
+
+    return pop;
+  }
+
+  /**
+   * The simulation should only end if the entire population is dead, everyone is vaccinated, or the virus died. 
+   * @return {boolean} - If simulation should continue. 
+   */
+  simulation_should_continue() {
+    // if no more infected people => end simulation since virus died
+    if (this.current_infected.length === 0) {
+      // console.log("virus ended");
+      return false;
+    }
+
+    for (let i = 0; i < this.population.length; i++) {
```

Comment:
maybe use `this.population.some()` here?

Reply:
I changed "this.population.length" to "this.pop_size".

Reply:
Changed to some() method